Remove the Legacy REST API from WooCommerce core

Konamiman opened the pull request:
### Changes proposed in this Pull Request:

This pull request implements the removal of the Legacy REST API from WooCommerce core, [as announced in the WooCommerce developers blog](https://developer.woocommerce.com/2023/10/03/the-legacy-rest-api-will-move-to-a-dedicated-extension-in-woocommerce-9-0/).

#### Endpoint changes

Any request for any endpoint whose URL starts with `/wc-api/v3` will now return a 404 error with the following contents, exactly in the same way as when the Legacy REST API was disabled in WooCommerce 8.9 and older:

```json
{
    "errors": {
        "code": "woocommerce_api_disabled",
        "message": "The WooCommerce API is disabled on this site"
    }
}
```

For compatibility, the Legacy REST API will be considered disabled (endpoints won't work) if the `woocommerce_api_enabled` is not set to `yes`, even if the plugin is active.

#### Changes in the Legacy API settings page

The WooCommerce - Advanced - Legacy API settings page will now tell whether the plugin is installed or not, but it won't be possible to actually change anything from that page:

![image](https://github.com/woocommerce/woocommerce/assets/937723/18677614-cc3c-4e46-8b63-16ee263a20ed)

![image](https://github.com/woocommerce/woocommerce/assets/937723/264306de-1f9b-4c06-8b23-4d1b544e0ddd)

#### Changes in the webhooks page

If one or more webhooks exist that are configured for using the Legacy REST API for composing their payloads and the Legacy REST API plugin is **not** active, a warning notice is shown in the  WooCommerce - Advanced - Webhooks page and a warning icon with a tooltip prepends the name of the "offending" webhooks:

![image](https://github.com/woocommerce/woocommerce/assets/937723/71a3d3f8-6c26-4549-aa5b-047da89be3a6)

Also, the wording of the already existing global admin notice about these webhooks has changed from "...will be unsupported soon" to "...is unsupported".

![image](https://github.com/woocommerce/woocommerce/assets/937723/0fbc7882-e990-4c09-876f-1a14e0fcc6c4)

Additionally, a "NOT AVAILABLE" warning will be displayed in the format dropdown in the webhook edit page when the legacy API payload format is selected (if the Legacy API plugin is installed, the warning text is "deprecated"):

![image](https://github.com/woocommerce/woocommerce/assets/937723/91f29725-77de-480f-b354-9152bb27fbb6)

![image](https://github.com/woocommerce/woocommerce/assets/937723/4561274e-6aa9-43eb-90d4-f53bc23b2462)

#### Changes in the status report

The WooCommerce status report (WooCommerce - Status - System status) will display the version of the installed Legacy REST API plugin, or a message indicating that the plugin is not installed:

![image](https://github.com/woocommerce/woocommerce/assets/937723/19d7bbea-305c-41be-b07e-ebbd8e20e584)

![image](https://github.com/woocommerce/woocommerce/assets/937723/6e7e6104-dd3e-45ea-af85-368bb817fa3e)

### How to test the changes in this Pull Request:

Test upgrading from an older WooCommerce to a build containing this pull request, and verifying that the Legacy REST API works or not depending on all the sensible combinations (the first three columns refer to the WooCommerce state _before_ the upgrade, the last two are _after_ the upgrade):

Legacy REST API active | Legacy webhook exists | Plugin was already installed | Plugin gets autoinstalled | Legacy REST API works
|-|-|-|-|-|
❌ | ❌ | ❌ | ❌ | ❌
❌ | ❌ | ✔️ | ❌ | ✔️
✔️ | ❌ | ✔️ | ❌ | ✔️
❌ | ✔️ | ✔️ | ❌ | ✔️
✔️ | ✔️ | ✔️ | ❌ | ✔️
✔️ | ❌ | ❌ | ✔️ | ✔️
❌ | ✔️ | ❌ | ✔️ | ✔️
✔️ | ✔️ | ❌ | ✔️ | ✔️

Notes:

* If the plugin gets autoinstalled, you'll get an admin notice and a message next to the plugin name in the plugins page (see https://github.com/woocommerce/woocommerce/pull/45570)
* "Legacy REST API works" means that you get the expected result when querying any of the `/wc-api/v3/` entpoints (e.g. `/wc-api/v3/orders`); the opposite means that you get the "The WooCommerce API is disabled on this site" message.
* If you are switching back to an older version of WooCommerce as part of your testing, remember that the value of the `wc_version` option won't be downgraded: you need to do it manually (e.g. `wc option set 8.7.0`) or the WooCommerce upgrade code won't run the next time you upgrade again to a higher version!

There's an extra test to perform. The store API was using `wc()->api->get_endpoint_data` to get the current cart under certain error conditions when hitting the checkout endpoint, this has been replaced with `wc_get_container()->get( RestApiUtil::class )->get_endpoint_data` which is equivalent. The easiest way to test this is to perform a temporary modification in the `Automattic\WooCommerce\StoreApi\Routes\V1\Checkout` class, rewriting the `get_response` method as follows:

```php
public function get_response( \WP_REST_Request $request ) {
     $this->load_cart_session( $request );
     $this->cart_controller->calculate_totals();
     $error = new \WP_Error('an_error', 'This is an error');
     $this->add_data_to_error_object($error, ['some_data' => 'hello'], 400, true);
     $error = $this->error_to_response( $error );
     return $this->add_response_headers( $error );
}
```

With this change, run `curl http://localhost/wp-json/wc/store/v1/checkout` and verify that the response contains information for an empty cart.

Konamiman committed:
Replace the Legacy REST API with a stub that always returns an error.

If the Legacy REST API plugin is installed, it will "take over"
the stub and the API will be available again without noticeable
changes in behavior.

Konamiman committed:
Merge branch 'trunk' into remove_legacy_rest_api

Konamiman committed:
Fix: required argument not passed to query_vars filter

Konamiman committed:
Fix code styling

Konamiman committed:
Move the get_endpoint_data method to WooCommerce core

This method was part of the legacy REST API package but is actually
being used to query endpoints of the new REST API, so it actually
belongs to WooCommerce core.

It has been placed in a new src/Utilities/RestApiUtil class.
The existing src/Internal/RestApiUtil class has been renamed
ro RestApiParameterUtil to avoid confussion.

Konamiman committed:
Fix the status report to account for the legacy REST API being missing.

Konamiman committed:
Move the register_wp_admin_settings to WooCommerce core.

This method was part of the legacy REST API package but it's needed
for certain endpoints of the current REST API, for example
settings/general/woocommerce_specific_allowed_countries

Konamiman committed:
Adjust unit tests for the removal of the legacy REST API.

Basically, some tests no longer make sense so they are removed.

Konamiman committed:
Adjust webhooks for the removal of the legacy REST API.

If a webhook is configured to compose its payload using the
legacy REST API format, and the separate legacy REST API plugin
is not installed, an exception will be thrown. This will cause
the scheduled action that does the webhook delivery to fail.

Konamiman committed:
Add warning in the webhooks settings page.

The warning will show if the legacy REST API is not available
and there's at least one plugin that is configured to be delivered
using legacy payloads. Each hook will have an exclamation symbol
next to its name in the table.

Additionally, inside the webhook edition page the legacy payload
option will display "NOT AVAILABLE" instead of "deprecated"
when the legacy REST API is not available.

Konamiman committed:
Add support for queries containing the "wc-api" variable.

These queries will return a 400 error by default, but
"woocommerce_api_request" and "woocommerce_api_(query)" actions
are available for custom processing.

Konamiman committed:
Merge branch 'trunk' into remove_legacy_rest_api

Konamiman committed:
Merge branch 'trunk' into remove_legacy_rest_api

Konamiman committed:
Merge branch 'trunk' into remove_legacy_rest_api

Konamiman committed:
Adjustments in messages, including adding links.

Also the status report page now displays the plugin version.

Konamiman committed:
Adjust the wording of the legacy webhooks incompatibility notice.

Also remove the "legacy API will be removed soon" notice.

Konamiman committed:
Add changelog file

Konamiman committed:
Enable the woocommerce_api_enabled option if the Legacy API plugin
is installed and the option is disabled.

Konamiman committed:
Fix linting issues

Konamiman committed:
Remove a test that is no longer relevant

coreymckrill commented on plugins/woocommerce/includes/admin/class-wc-admin-webhooks.php:
```diff
@@ -323,6 +325,81 @@ private static function table_list_output() {
+						/* translators: %s = URL */
+						_n(
+							'This webhook has the ⚠ symbol in front of its name in the list below. Please either edit the webhook to use a different delivery format, or install and activate <a href="%s" target="_blank">the WooCommerce Legacy REST API extension</a>.',
+							'These webhooks have the ⚠ symbol in front of their names in the list below. Please either edit the webhooks to use a different delivery format, or install and activate <a href="%s" target="_blank">the WooCommerce Legacy REST API extension</a>.',
```
```suggestion
							'This webhook has the ⚠️ symbol in front of its name in the list below. Please either edit the webhook to use a different delivery format, or install and activate <a href="%s" target="_blank">the WooCommerce Legacy REST API extension</a>.',
							'These webhooks have the ⚠️ symbol in front of their names in the list below. Please either edit the webhooks to use a different delivery format, or install and activate <a href="%s" target="_blank">the WooCommerce Legacy REST API extension</a>.',
```

coreymckrill commented on plugins/woocommerce/includes/admin/class-wc-admin-webhooks-table-list.php:
```diff
@@ -75,7 +75,13 @@ public function column_title( $webhook ) {
+		$warning_prefix =
+			is_null( wc()->api ) && $this->uses_legacy_rest_api( $webhook ) ?
+			sprintf(
+				"<span title='%s'>⚠</span>️ ",
```
```suggestion
				"<span title='%s'>⚠️</span>️ ",
```

coreymckrill commented on plugins/woocommerce/includes/admin/class-wc-admin-notices.php:
```diff
@@ -159,51 +159,23 @@ public static function reset_admin_notices() {
 					sprintf(
 					// translators: Placeholders are URLs.
-						wpautop( __( 'The WooCommerce Legacy REST API will be removed in WooCommerce 9.0, and this will cause <a href="%1$s">webhooks on this site that are configured to use the Legacy REST API</a> to stop working. <a target="_blank" href="%2$s">A separate WooCommerce extension is available</a> to allow these webhooks to keep using the Legacy REST API without interruption. You can also edit these w ...
+						wpautop( __( '⚠ The WooCommerce Legacy REST API has been removed from WooCommerce, this will cause <a href="%1$s">webhooks on this site that are configured to use the Legacy REST API</a> to stop working. <a target="_blank" href="%2$s">A separate WooCommerce extension is available</a> to allow these webhooks to keep using the Legacy REST API without interruption. You can also edit these webh ...
```
```suggestion
						wpautop( __( '⚠️ The WooCommerce Legacy REST API has been removed from WooCommerce, this will cause <a href="%1$s">webhooks on this site that are configured to use the Legacy REST API</a> to stop working. <a target="_blank" href="%2$s">A separate WooCommerce extension is available</a> to allow these webhooks to keep using the Legacy REST API without interruption. You can also edit these webhooks to use the current REST API version to generate the payload instead. <b><a target="_blank" href="%3$s">Learn more about this change.</a></b>', 'woocommerce' ) ),
```

coreymckrill commented on plugins/woocommerce/includes/admin/settings/views/html-webhooks-edit.php:
```diff
@@ -145,7 +145,13 @@
+						<?php
+						$legacy_api_option_name =
+							is_null( wc()->api ) ?
+							__( 'Legacy API v3 (⚠ NOT AVAILABLE)', 'woocommerce' ) :
```
```suggestion
							__( 'Legacy API v3 (⚠️ NOT AVAILABLE)', 'woocommerce' ) :
```

coreymckrill commented on plugins/woocommerce/includes/class-woocommerce.php:
```diff
@@ -1175,6 +1177,160 @@ public function get_global( string $global_name ) {
 		return wc_get_container()->get( LegacyProxy::class )->get_global( $global_name );
 	}
 
```
🤔 I'm wondering if it might make sense to keep all this code for setting up the stub in a separate file (maybe in the `Internal` namespace, even?). Seems like it adds a lot of clutter to the main, public WooCommerce class for something that will increasingly be an irrelevant edge case.

Konamiman replied:
That's a very good idea! I have moved the stub to a dedicated internal class, `LegacyRestApiStub` in `src/Internal/Utilities`.

coreymckrill reviewed (commented):
Mostly a bunch of suggestions about standardizing around ⚠️ instead of having both that and ⚠.

Konamiman committed:
Standardize exclamation sign characters in notices

Co-authored-by: Corey McKrill <916023+coreymckrill@users.noreply.github.com>

Konamiman committed:
Merge branch 'trunk' into remove_legacy_rest_api

Konamiman committed:
Move legacy API stub initialization to a separate class

jorgeatorres commented on plugins/woocommerce/includes/class-woocommerce.php:
```diff
@@ -254,6 +258,7 @@ private function init_hooks() {
 		add_action( 'deactivated_plugin', array( $this, 'deactivated_plugin' ) );
 		add_action( 'woocommerce_installed', array( $this, 'add_woocommerce_inbox_variant' ) );
 		add_action( 'woocommerce_updated', array( $this, 'add_woocommerce_inbox_variant' ) );
+		self::add_action( 'woocommerce_init', array( $this, 'register_wp_admin_settings' ) );
```
Prior to this PR, this was being done [in `rest_api_init`](https://github.com/woocommerce/woocommerce/blob/ace978691401af09bea34dd0ee716a49822ef4e3/plugins/woocommerce/includes/class-wc-api.php#L29), not `woocommerce_init`. Unless we have good reason to change that, I guess it'd be more compatible to use the same hook as before.

Konamiman replied:
I can't figure out why this was changed. I have reverted it back to `rest_api_init`.

jorgeatorres commented on plugins/woocommerce/includes/wc-core-functions.php:
```diff
@@ -1102,20 +1102,19 @@ function wc_setcookie( $name, $value, $expire = 0, $secure = false, $httponly =
  * @param string $path an endpoint to include in the URL.
  * @return string the URL.
  */
 function get_woocommerce_api_url( $path ) {
```
Should we explicitly mark this as deprecated with a `@deprecated` docblock tag?

Also, there's a unit test (`WC_Tests_Core_Functions::test_get_woocommerce_api_url`) around this method. Should that be removed too?

Konamiman replied:
Good idea, deprecation tag added.

As for the unit test, it doesn't hurt to keep it in place, also I had modified it too so that it assumes version is `v3`, as this value is now hardcoded in the function.

jorgeatorres commented on plugins/woocommerce/includes/admin/views/html-admin-page-status-report.php:
```diff
@@ -96,16 +97,16 @@
 			<td>
 				<?php
-				$version = wc()->api->get_rest_api_package_version();
+				$version = is_null( wc()->api ) ? null : get_plugin_data( WP_PLUGIN_DIR . '/woocommerce-legacy-rest-api/woocommerce-legacy-rest-api.php' )['Version'] ?? '';
```
This check is not robust enough. The Legacy REST API plugin might be active but using a different name or located in a different plugin directory (for example, as a mu-plugin or similar). In that case, this will produce a PHP warning.

To test, just rename the legacy REST API plugin directory and enable it.

Konamiman replied:
Good catch. I have changed it so that it now uses `PluginUtil::get_wp_plugin_id` to get the proper path of the plugin.

jorgeatorres commented on plugins/woocommerce/src/Internal/RestApiParameterUtil.php:
```diff
@@ -6,13 +6,13 @@
  * @package Automattic\WooCommerce\Internal
  */
-class RestApiUtil {
+class RestApiParameterUtil {
```
It doesn't seem like this was used elsewhere, and while it is inside the "internal" namespace, it has existed for a few years now. Probably nothing will break, but not sure if we _really_ need to rename it / change it right now. Thoughts?

Konamiman replied:
I think that the point of having the `Internal` namespace is precisely to have freedom to perform this kind of refactorings. This class was indeed used but by us, internally, and this PR renames all the usages too (and introduces a different class named `RestApiUtil`).

jorgeatorres reviewed (changes requested):
Hi @Konamiman!

Overall this looks good and testing went well. Thanks for all the work you put into this! 🥇 
I left some inline comments, but nothing too serious.

I'd also like to point out a couple of things:

- Should we add something about the deprecation to the `$api` var in the main WC class? Right now, it seems like a var that is "never used".
- The StoreAPI seems to still be using `wc()->api->...` [here](https://github.com/woocommerce/woocommerce/blob/afee75f87110555e2647dd344efa9a37a1e73a69/plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php#L324). I haven't followed exactly when or how this gets used, but it'd be good to confirm it's not a problem.

Please let me know what you think.

Konamiman committed:
Run register_wp_admin_settings in rest_api_init hook instead of woocommerce_init

Konamiman committed:
Mark get_woocommerce_api_url function as deprecated

Konamiman committed:
More robust check for the legacy API plugin version check in status report

Konamiman committed:
Mark $api property in WooCommerce class as deprecated

Konamiman committed:
Replace call to wc()->api->get_endpoint_data in Checkout class

Replace it with just wc()->cart, which is what the invoked
endpoint was doing in the end.

Konamiman committed:
Merge branch 'trunk' into remove_legacy_rest_api

Konamiman commented:
> The StoreAPI seems to still be using wc()->api->...

Good one. This pull request was created seven months ago and while I took care of finding all the usages of `wc()->api` in the codebase, this one escaped because back then Blocks (and thus the store API) was still not part of core.

Fortunately the fix is easy. What `wc()->api->get_endpoint_data` does is to programmatically invoke and endpoint, and what `/wc/store/v1/cart` does is simply to retrieve `wc()->cart`. So I've updated the code to do exactly that.

Konamiman committed:
Use get_endpoint_data to get the data from /wc/store/v1/cart

jorgeatorres reviewed (approved):
LGTM. I believe there are still some linting issues with the JS and I see some on the PHP side (at least [here](https://github.com/woocommerce/woocommerce/blob/9e15e11a5c42004989262ea14851aa3ed56b144b/plugins/woocommerce/includes/admin/views/html-admin-page-status-report.php#L108)), but for some reason are not being reported.

Could you please look into that @Konamiman?

Konamiman committed:
Additional beautification of html-admin-page-status-report.php

Konamiman committed:
Merge branch 'trunk' into remove_legacy_rest_api

Konamiman committed:
Fix linting issues in settings-crud.test.js

Konamiman committed:
Yet more beautification of html-admin-page-status-report.php

Konamiman commented:
Given the significance of this pull request I've added the `needs: analysis` label, but given that it needs to go into WooCommerce 9.0 no matter what, I've merged it already. If the analysis finds issues these will be handled in separate pull request(s), and code freeze exceptions for WooCommerce 9.0 will be created if necessary.
